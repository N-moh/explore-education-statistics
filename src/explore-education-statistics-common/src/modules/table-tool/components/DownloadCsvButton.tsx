import ButtonText from '@common/components/ButtonText';
import cartesian from '@common/lib/utils/cartesian';
import {
  CategoryFilter,
  LocationFilter,
  TimePeriodFilter,
  Filter,
} from '@common/modules/full-table/types/filters';
import { saveAs } from 'file-saver';
import Papa from 'papaparse';
import React from 'react';
import { transformTableMetaFiltersToCategoryFilters } from '@common/modules/full-table/utils/tableHeaders';
import { FullTable } from '@common/modules/full-table/types/fullTable';

interface Props {
  publicationSlug: string;
  fullTable: FullTable;
}

const DownloadCsvButton = ({ publicationSlug, fullTable }: Props) => {
  const { subjectMeta, results } = fullTable;
  const {
    indicators,
    filters: metaFilters,
    timePeriodRange: timePeriods,
    locations,
  } = subjectMeta;

  const filters = transformTableMetaFiltersToCategoryFilters(metaFilters);

  const getCsvData = (): string[][] => {
    const filterColumns = Object.entries(metaFilters).map(
      ([key]) => metaFilters[key].legend,
    );

    const indicatorColumns = indicators.map(indicator => {
      const unit = indicator.unit ? ` (${indicator.unit})` : '';
      return `${indicator.label}${unit}`;
    });

    const columns = [
      'Location',
      'Time period',
      ...filterColumns,
      ...indicatorColumns,
    ];

    const rows = cartesian<Filter>(
      locations,
      timePeriods,
      ...Object.values(filters),
    ).map(row => {
      // TODO: Remove ignore when Prettier stops adding trailing comma to tuple type
      // prettier-ignore
      const [location, timePeriod, ...filterOptions] = row as [
        LocationFilter,
        TimePeriodFilter,
        ...CategoryFilter[]
      ];

      const indicatorCells = indicators.map(indicator => {
        const matchingResult = results.find(result => {
          return Boolean(
            filterOptions.every(filter =>
              result.filters.includes(filter.value),
            ) &&
              result.timePeriod === timePeriod.value &&
              result.location[location.level] &&
              result.location[location.level].code === location.value,
          );
        });
        if (!matchingResult) {
          return 'n/a';
        }
        return matchingResult.measures[indicator.value];
      });

      return [...row.map(column => column.label), ...indicatorCells];
    });

    return [columns, ...rows];
  };

  return (
    <ButtonText
      onClick={() => {
        const data = new Blob([Papa.unparse(getCsvData())], {
          type: 'text/csv;charset=utf-8',
        });

        const filename = `data-${publicationSlug}.csv`;

        saveAs(data, filename, {
          autoBom: true,
        });
      }}
    >
      Download the underlying data of this table (.CSV)
    </ButtonText>
  );
};

export default DownloadCsvButton;
