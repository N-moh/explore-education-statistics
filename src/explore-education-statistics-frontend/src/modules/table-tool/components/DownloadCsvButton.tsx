import ButtonText from '@common/components/ButtonText';
import cartesian from '@common/lib/utils/cartesian';
import {
  PublicationSubjectMeta,
  TableData,
} from '@common/services/tableBuilderService';
import { Dictionary } from '@common/types';
import {
  CategoryFilter,
  Indicator,
  LocationFilter,
  Filter,
} from '@frontend/modules/table-tool/components/types/filters';
import TimePeriod from '@frontend/modules/table-tool/components/types/TimePeriod';
import { saveAs } from 'file-saver';
import Papa from 'papaparse';
import React from 'react';

interface Props {
  publicationSlug: string;
  meta: PublicationSubjectMeta;
  indicators: Indicator[];
  filters: Dictionary<CategoryFilter[]>;
  timePeriods: TimePeriod[];
  locations: LocationFilter[];
  results: TableData['results'];
}

const DownloadCsvButton = ({
  publicationSlug,
  meta,
  indicators,
  filters,
  timePeriods,
  locations,
  results,
}: Props) => {
  const getCsvData = (): string[][] => {
    const filterColumns = Object.entries(filters).map(
      ([key]) => meta.filters[key].legend,
    );

    const indicatorColumns = indicators.map(indicator => {
      const unit = indicator.unit ? ` (${indicator.unit})` : '';
      return `${indicator.label}${unit}`;
    });

    const columns = [
      'Location',
      'Time period',
      ...filterColumns,
      ...indicatorColumns,
    ];

    const rows = cartesian<Filter>(
      locations,
      timePeriods,
      ...Object.values(filters),
    ).map(row => {
      // TODO: Remove ignore when Prettier stops adding trailing comma to tuple type
      // prettier-ignore
      const [location, timePeriod, ...filterOptions] = row as [
        LocationFilter,
        TimePeriod,
        ...CategoryFilter[]
      ];

      const indicatorCells = indicators.map(indicator => {
        const matchingResult = results.find(result => {
          return Boolean(
            filterOptions.every(filter =>
              result.filters.includes(filter.value),
            ) &&
              result.timePeriod === timePeriod.value &&
              result.location[location.level] &&
              result.location[location.level].code === location.value,
          );
        });

        if (!matchingResult) {
          return 'n/a';
        }

        return matchingResult.measures[indicator.value];
      });

      return [...row.map(column => column.label), ...indicatorCells];
    });

    return [columns, ...rows];
  };

  return (
    <ButtonText
      onClick={() => {
        const data = new Blob([Papa.unparse(getCsvData())], {
          type: 'text/csv;charset=utf-8',
        });

        const filename = `data-${publicationSlug}.csv`;

        saveAs(data, filename, {
          autoBom: true,
        });
      }}
    >
      Download underlying data (.csv)
    </ButtonText>
  );
};

export default DownloadCsvButton;
