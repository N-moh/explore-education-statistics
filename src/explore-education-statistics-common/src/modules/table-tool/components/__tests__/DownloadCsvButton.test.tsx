import {
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { writeFile } from 'xlsx';
import DownloadCsvButton, { getCsvData } from '../DownloadCsvButton';

jest.mock('xlsx', () => {
  const { utils } = jest.requireActual('xlsx');

  return {
    writeFile: jest.fn(),
    utils,
  };
});

describe('DownloadCsvButton', () => {
  const basicSubjectMeta = {
    geoJsonAvailable: false,
    publicationName: '',
    subjectName: '',
    footnotes: [],
    filters: {
      characteristics: {
        legend: 'Characteristics',
        hint: '',
        options: {
          gender: {
            label: 'Gender',
            options: [{ label: 'Female', value: 'gender_female' }],
          },
        },
      },
    },
    indicators: [
      new Indicator({
        label: 'Authorised absence rate',
        value: 'authAbsRate',
        unit: '%',
      }),
      new Indicator({
        label: 'Number of authorised absence sessions',
        value: 'authAbsSess',
        unit: '',
      }),
    ],
    locations: [
      new LocationFilter(
        {
          value: 'england',
          label: 'England',
        },
        'country',
      ),
    ],
    timePeriodRange: [
      new TimePeriodFilter({ code: 'AY', year: 2015, label: '2015/16' }),
    ],
  };

  test('calls `writeFile` when button is pressed', () => {
    const { getByText } = render(
      <DownloadCsvButton
        publicationSlug="pupil-absence"
        fullTable={{
          subjectMeta: {
            ...basicSubjectMeta,
            filters: {
              ...basicSubjectMeta.filters,
              schoolType: {
                legend: 'School type',
                hint: '',
                options: {
                  default: {
                    label: '',
                    options: [
                      {
                        label: 'State-funded-primary',
                        value: 'school_primary',
                      },
                      {
                        label: 'State-funded secondary',
                        value: 'school_secondary',
                      },
                    ],
                  },
                },
              },
            },
          },
          results: [
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2015_AY',
              geographicLevel: 'Country',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '123',
                authAbsSess: '456',
              },
            },
            {
              filters: ['gender_female', 'school_secondary'],
              timePeriod: '2015_AY',
              geographicLevel: 'Country',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '234',
                authAbsSess: '567',
              },
            },
          ],
        }}
      />,
    );

    fireEvent.click(
      getByText('Download the underlying data of this table (CSV)'),
    );

    const mockedWriteFile = writeFile as jest.Mock;

    expect(mockedWriteFile).toHaveBeenCalledTimes(1);
    expect(mockedWriteFile.mock.calls[0][0]).toMatchSnapshot();
    expect(mockedWriteFile.mock.calls[0][1]).toBe('data-pupil-absence.csv');
  });

  describe('getCsvData', () => {
    test('contains full set of data', async () => {
      const data = getCsvData({
        subjectMeta: {
          ...basicSubjectMeta,
          filters: {
            ...basicSubjectMeta.filters,
            schoolType: {
              legend: 'School type',
              hint: '',
              options: {
                default: {
                  label: '',
                  options: [
                    {
                      label: 'State-funded-primary',
                      value: 'school_primary',
                    },
                    {
                      label: 'State-funded secondary',
                      value: 'school_secondary',
                    },
                  ],
                },
              },
            },
          },
        },
        results: [
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '123',
              authAbsSess: '456',
            },
          },
          {
            filters: ['gender_female', 'school_secondary'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '234',
              authAbsSess: '567',
            },
          },
        ],
      });

      expect(data).toHaveLength(3);

      expect(data[0]).toHaveLength(6);

      expect(data[1]).toHaveLength(6);
      expect(data[1][4]).toBe('123');
      expect(data[1][5]).toBe('456');

      expect(data[2]).toHaveLength(6);
      expect(data[2][4]).toBe('234');
      expect(data[2][5]).toBe('567');

      expect(data).toMatchSnapshot();
    });

    test("contains n/a's if there are some matching results", () => {
      const data = getCsvData({
        subjectMeta: {
          ...basicSubjectMeta,
          filters: {
            ...basicSubjectMeta.filters,
            schoolType: {
              legend: 'School type',
              hint: '',
              options: {
                default: {
                  label: '',
                  options: [
                    {
                      label: 'State-funded-primary',
                      value: 'school_primary',
                    },
                    {
                      label: 'State-funded secondary',
                      value: 'school_secondary',
                    },
                  ],
                },
              },
            },
          },
        },
        results: [
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '123',
              authAbsSess: '456',
            },
          },
        ],
      });

      expect(data).toHaveLength(3);

      expect(data[0]).toHaveLength(6);

      expect(data[1]).toHaveLength(6);
      expect(data[1][4]).toBe('123');
      expect(data[1][5]).toBe('456');

      expect(data[2]).toHaveLength(6);
      expect(data[2][4]).toBe('n/a');
      expect(data[2][5]).toBe('n/a');

      expect(data).toMatchSnapshot();
    });

    test("contains n/a's if there are no results", () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(5);

      expect(data[1]).toHaveLength(5);
      expect(data[1][3]).toBe('n/a');
      expect(data[1][4]).toBe('n/a');

      expect(data).toMatchSnapshot();
    });

    test("contains n/a's if matching result but no matching indicator", () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsSess: '123',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(5);

      expect(data[1]).toHaveLength(5);
      expect(data[1][3]).toBe('n/a');
      expect(data[1][4]).toBe('123');

      expect(data).toMatchSnapshot();
    });

    test('does not format values in any way', () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '12300000',
              authAbsSess: '44255667.2356',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(5);

      expect(data[1]).toHaveLength(5);
      expect(data[1][3]).toBe('12300000');
      expect(data[1][4]).toBe('44255667.2356');

      expect(data).toMatchSnapshot();
    });

    test('can contain suppressed cells', () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'Country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '13.4',
              authAbsSess: 'x',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(5);

      expect(data[1]).toHaveLength(5);
      expect(data[1][3]).toBe('13.4');
      expect(data[1][4]).toBe('x');

      expect(data).toMatchSnapshot();
    });
  });
});
