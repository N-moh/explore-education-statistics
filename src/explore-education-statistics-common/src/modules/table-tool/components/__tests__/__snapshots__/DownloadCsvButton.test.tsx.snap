// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCsvButton calls \`writeFile\` when button is pressed 1`] = `
Object {
  "SheetNames": Array [
    "Sheet1",
  ],
  "Sheets": Object {
    "Sheet1": Object {
      "!ref": "A1:G2",
      "A1": Object {
        "t": "s",
        "v": "location",
      },
      "A2": Object {
        "t": "s",
        "v": "England",
      },
      "B1": Object {
        "t": "s",
        "v": "location_code",
      },
      "B2": Object {
        "t": "s",
        "v": "england",
      },
      "C1": Object {
        "t": "s",
        "v": "geographic_level",
      },
      "C2": Object {
        "t": "s",
        "v": "country",
      },
      "D1": Object {
        "t": "s",
        "v": "time_period",
      },
      "D2": Object {
        "t": "s",
        "v": "201516",
      },
      "E1": Object {
        "t": "s",
        "v": "characteristic",
      },
      "E2": Object {
        "t": "s",
        "v": "Female",
      },
      "F1": Object {
        "t": "s",
        "v": "sess_authorised_percent",
      },
      "F2": Object {
        "t": "s",
        "v": "n/a",
      },
      "G1": Object {
        "t": "s",
        "v": "sess_authorised",
      },
      "G2": Object {
        "t": "s",
        "v": "n/a",
      },
    },
  },
}
`;

exports[`DownloadCsvButton getCsvData can contain suppressed cells 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "13.4",
    "x",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains full set of data 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "school_type",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded primary",
    "111",
    "222",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded secondary",
    "333",
    "444",
  ],
  Array [
    "Barnsley",
    "barnsley",
    "localAuthority",
    "201516",
    "Female",
    "State-funded primary",
    "555",
    "666",
  ],
  Array [
    "Barnsley",
    "barnsley",
    "localAuthority",
    "201516",
    "Female",
    "State-funded secondary",
    "777",
    "888",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if matching result but no matching indicator 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "n/a",
    "111",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if there are no results 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "n/a",
    "n/a",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if there are some matching results 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "school_type",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded primary",
    "111",
    "222",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded secondary",
    "n/a",
    "n/a",
  ],
]
`;

exports[`DownloadCsvButton getCsvData does not format values in any way 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "12300000",
    "44255667.2356",
  ],
]
`;
